import os

from ament_index_python.packages import get_package_share_directory # type: ignore
from launch import LaunchDescription
from launch.actions import DeclareLaunchArgument, GroupAction, SetEnvironmentVariable, IncludeLaunchDescription # type: ignore
from launch.conditions import IfCondition # type: ignore
from launch.substitutions import LaunchConfiguration, PythonExpression # type: ignore
from launch_ros.actions import LoadComposableNodes # type: ignore
from launch_ros.actions import Node # type: ignore
from launch_ros.descriptions import ComposableNode, ParameterFile # type: ignore
from nav2_common.launch import RewrittenYaml # type: ignore
from launch.launch_description_sources import PythonLaunchDescriptionSource # type: ignore
from launch.substitutions import LaunchConfiguration, PythonExpression # type: ignore


def generate_launch_description():
    # Get the package directory
    pkg_dir = get_package_share_directory('navigation2_run')

    # Launch configurations
    namespace = LaunchConfiguration('namespace')
    map_yaml_file = LaunchConfiguration('map')
    use_sim_time = LaunchConfiguration('use_sim_time')
    autostart = LaunchConfiguration('autostart')
    params_file = LaunchConfiguration('params_file')
    use_composition = LaunchConfiguration('use_composition')
    use_respawn = LaunchConfiguration('use_respawn')
    log_level = LaunchConfiguration('log_level')
    use_odometry_sim = LaunchConfiguration('use_odometry_sim')
    robot_pose_remap = LaunchConfiguration('robot_pose_remap')

    lifecycle_nodes = ['map_server']

    # Remappings
    remappings = [('/tf', 'tf'),
                  ('/tf_static', 'tf_static'),
                  ('/odom/wheel', robot_pose_remap)]

    # Parameter substitution
    param_substitutions = {
        'use_sim_time': use_sim_time,
        'yaml_filename': map_yaml_file}

    configured_params = ParameterFile(
        RewrittenYaml(
            source_file=params_file,
            root_key=namespace,
            param_rewrites=param_substitutions,
            convert_types=True),
        allow_substs=True)

    # Environment variables
    stdout_linebuf_envvar = SetEnvironmentVariable(
        'RCUTILS_LOGGING_BUFFERED_STREAM', '1')

    # Declare launch arguments
    declare_namespace_cmd = DeclareLaunchArgument(
        'namespace', default_value='', description='Top-level namespace')
    declare_map_yaml_cmd = DeclareLaunchArgument(
        'map', description='Full path to map yaml file to load')
    declare_use_sim_time_cmd = DeclareLaunchArgument(
        'use_sim_time', default_value='false', description='Use simulation (Gazebo) clock if true')
    declare_params_file_cmd = DeclareLaunchArgument(
        'params_file', default_value=os.path.join(pkg_dir, 'params', 'nav2_params.yaml'),
        description='Full path to the ROS2 parameters file to use for all launched nodes')
    declare_autostart_cmd = DeclareLaunchArgument(
        'autostart', default_value='true', description='Automatically startup the nav2 stack')
    declare_use_composition_cmd = DeclareLaunchArgument(
        'use_composition', default_value='False', description='Use composed bringup if True')
    declare_container_name_cmd = DeclareLaunchArgument(
        'container_name', default_value='nav2_container',
        description='The name of container that nodes will load in if use composition')
    declare_use_respawn_cmd = DeclareLaunchArgument(
        'use_respawn', default_value='False', description='Whether to respawn if a node crashes.')
    declare_log_level_cmd = DeclareLaunchArgument(
        'log_level', default_value='info', description='Log level')

    # Load static transform & odometry simulation
    # localization_sim = GroupAction(
    #     condition=IfCondition(use_odometry_sim),
    #     actions=[
    #         Node(
    #             package='navigation2_run',
    #             executable='odometry_sim',
    #             name='odometry_sim',
    #             output='screen',
    #             respawn=use_respawn,
    #             respawn_delay=2.0,
    #             parameters=[{"cmd_cb_name": "/cmd_vel"}],
    #             arguments=['--ros-args', '--log-level', log_level],
    #             remappings=remappings
    #         )
    #     ]
    # )


    # Bringup localization sim nodes
    # brinup_localization_cmd_group = GroupAction([
    #     IncludeLaunchDescription(
    #             PythonLaunchDescriptionSource(os.path.join(get_package_share_directory('sima-localization-sim'), 'launch', 'localization_sim.launch.py')),
    #             launch_arguments={'namespace': namespace,
    #                             'use_sim_time': use_sim_time,
    #                             'autostart': autostart,
    #                             'params_file': params_file,
    #                             'use_composition': use_composition,
    #                             'remappings': str(remappings),
    #                             'use_respawn': use_respawn}.items()),

    #     IncludeLaunchDescription(
    #         PythonLaunchDescriptionSource(os.path.join(get_package_share_directory('sima-localization-sim'), 'launch', 'robot_localization.launch.py')),
    #         launch_arguments={'namespace': namespace,
    #                             'use_sim_time': use_sim_time,
    #                             'autostart': autostart,
    #                             'params_file': params_file,
    #                             'use_composition': use_composition,
    #                             'use_respawn': use_respawn}.items()),
    # ])

    # for gazebo simulation, we just publish a static transform from map to odom
    # fake_localization_tf = Node(
    #     package='tf2_ros',
    #     executable='static_transform_publisher',
    #     name='map_to_odom_static_publisher',
    #     arguments=['0', '0', '0', '0', '0', '0', 'map', 'odom'],
    #     parameters=[{'use_sim_time': use_sim_time}],
    #     output='screen'
    # )


    # Load nodes group
    load_nodes = GroupAction(
        condition=IfCondition(PythonExpression(['not ', use_composition])),
        actions=[
            Node(
                package='nav2_map_server',
                executable='map_server',
                name='map_server',
                output='screen',
                respawn=use_respawn,
                respawn_delay=2.0,
                parameters=[configured_params],
                arguments=['--ros-args', '--log-level', log_level],
                remappings=remappings),
            Node(
                package='nav2_lifecycle_manager',
                executable='lifecycle_manager',
                name='lifecycle_manager_localization',
                output='screen',
                arguments=['--ros-args', '--log-level', log_level],
                parameters=[{'use_sim_time': use_sim_time},
                            {'autostart': autostart},
                            {'node_names': lifecycle_nodes}])
        ]
    )

    # Load composable nodes
    load_composable_nodes = LoadComposableNodes(
        condition=IfCondition(use_composition),
        target_container=LaunchConfiguration('container_name'),
        composable_node_descriptions=[
            ComposableNode(
                package='nav2_map_server',
                plugin='nav2_map_server::MapServer',
                name='map_server',
                parameters=[configured_params],
                remappings=remappings),
            ComposableNode(
                package='nav2_lifecycle_manager',
                plugin='nav2_lifecycle_manager::LifecycleManager',
                name='lifecycle_manager_localization',
                parameters=[{'use_sim_time': use_sim_time,
                             'autostart': autostart,
                             'node_names': lifecycle_nodes}]),
        ],
    )

    # Create the launch description
    ld = LaunchDescription()

    # Add environment variables
    ld.add_action(stdout_linebuf_envvar)

    # Add launch arguments
    ld.add_action(declare_namespace_cmd)
    ld.add_action(declare_map_yaml_cmd)
    ld.add_action(declare_use_sim_time_cmd)
    ld.add_action(declare_params_file_cmd)
    ld.add_action(declare_autostart_cmd)
    ld.add_action(declare_use_composition_cmd)
    ld.add_action(declare_container_name_cmd)
    ld.add_action(declare_use_respawn_cmd)
    ld.add_action(declare_log_level_cmd)

    # Add nodes to launch description
    # ld.add_action(localization_sim)
    ld.add_action(load_nodes)
    ld.add_action(load_composable_nodes)
    # ld.add_action(brinup_localization_cmd_group)      # for localization sim
    # ld.add_action(fake_localization_tf)               # for gazebo simulation

    return ld
